### ISSUE TITLE
Deleting deployment won't delete replicaset and associated pod

### ISSUE THREAD

**Issue by imyashvinder** [kubernetes/kubernetes#96362]
Hi,

I have recently noticed that I have lots of duplicated deployments because kubectl delete deployment won't delete the replica set and pod associated with it.

```
➜  ~ k get all
No resources found in deploymentbug namespace.
```
```
➜  ~ k create deploy nginx --image nginx                                                                                                                      
deployment.apps/nginx created
```
```
➜  ~ k get all
NAME                        READY   STATUS    RESTARTS   AGE
pod/nginx-f89759699-vvvbs   1/1     Running   0          9s

NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/nginx   1/1     1            1           9s

NAME                              DESIRED   CURRENT   READY   AGE
replicaset.apps/nginx-f89759699   1         1         1       9s
```
```
➜  ~ k delete deploy nginx
deployment.apps "nginx" deleted
```
```
➜  ~ k get all            
NAME                        READY   STATUS    RESTARTS   AGE
pod/nginx-f89759699-vvvbs   1/1     Running   0          22s

NAME                              DESIRED   CURRENT   READY   AGE
replicaset.apps/nginx-f89759699   1         1         1       22s
```

Kubernetes Version: 1.18
Platform: EKS
```➜  ~ k version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.9-eks-d1db3c", GitCommit:"d1db3c46e55f95d6a7d3e5578689371318f95ff9", GitTreeState:"clean", BuildDate:"2020-10-20T22:18:07Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}```

**Comment by pacoxu** [comment]
Can I know your kubectl version  and the k alias? 

```
|cascade |   | true | If true, cascade the deletion of the resources managed by this resource (e.g. Pods created by a ReplicationController). Default true.
-- | -- | -- | --

```
'kubectl delete' will use delete deployment with '--cascade true' by default.

![image](https://user-images.githubusercontent.com/2010320/98885443-fe1a5900-24cc-11eb-8af4-57c44239f0a2.png)

**Comment by cccs-eric** [comment]
Hi @pacoxu , I am also experiencing the same issue.  I created a [script](https://gist.github.com/cccs-eric/38c43465778f294176ce33deb484f7ca) that exposes the problem.  See if you can replicate on your side.  Here is the output of that script when I run it:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-16T20:43:08Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```
```
+ kubectl create namespace delete-test
namespace/delete-test created
```
```
+ kubectl -n delete-test create deploy nginx --image nginx
deployment.apps/nginx created
```
```
+ kubectl -n delete-test get all
NAME                         READY   STATUS    RESTARTS   AGE
pod/nginx-6799fc88d8-29zmh   1/1     Running   0          5s

NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/nginx   1/1     1            1           5s

NAME                               DESIRED   CURRENT   READY   AGE
replicaset.apps/nginx-6799fc88d8   1         1         1       5s
```
```
+ kubectl -n delete-test delete deploy nginx --cascade=true
deployment.apps "nginx" deleted
```
```
+ kubectl -n delete-test get all
NAME                         READY   STATUS    RESTARTS   AGE
pod/nginx-6799fc88d8-29zmh   1/1     Running   0          11s

NAME                               DESIRED   CURRENT   READY   AGE
replicaset.apps/nginx-6799fc88d8   1         1         1       11s
```
```
+ kubectl delete namespace delete-test
namespace "delete-test" deleted
```

**Comment by imyashvinder** [comment]
@cccs-eric This has been resolved for me. Actually, we are using EKS and it turned out to be a webhook issue.

I then deleted the webhook and applied again as below:

kubectl delete -f https://github.com/jetstack/cert-manager/releases/download/v0.11.1/cert-manager.yaml
kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.11.1/cert-manager.yaml

**Comment by cccs-eric** [comment]
@imyashvinder I'm convinced that it must be caused by something external since I can't believe such an obvious bug would make it out to a release.  I just can't figure out what is causing this.  I am on Azure AKS.  I'm not familiar with webhooks, so may I ask how you came up to this conclusion in your case?

**Comment by imyashvinder** [comment]
@cccs-eric We just tried replicating the issue and had a close look at api server and controller manager logs and found webhook related failure.
Tried running pod on specific nodes. Even created new aws managed node group, futile effort. Eventually removed the webhook and voila !!

**Comment by cccs-eric** [comment]
@imyashvinder I just tried on two other different AKS clusters (both running 1.18.8) and the delete works.  So it seems related to that one cluster.  I'll have to find what's different, other than the k8s version.

Thanks for the feedback 👍

**Comment by imyashvinder** [comment]
@cccs-eric why don't you raise a support request and let 'em have a look at it.

**Comment by cccs-eric** [comment]
I tested it on minikube 1.19.3 and 1.19.4 and it works on both.  So I'll raise it to Azure.

**Comment by cccs-eric** [comment]
@imyashvinder So after 3 days of trying to figure out what is going on, I should have listened to you...  Long story short, I was trying to install Eclipse Che and it comes with cert-manager unless you disable that option to use your own (which I do since I run Bitnami BKPR).  I did install Che once or twice with the cert-manager option set to true until I realized what was happening.  I guess it has corrupted in some way the BKPR cert-manager install.  I went into the kubelet and kube-controller-manager logs and saw some errors related to a missing service:
```
Failed to watch *v1.PartialObjectMetadata: failed to list *v1.PartialObjectMetadata: conversion webhook for cert-manager.io/v1alpha2, Kind=ClusterIssuer failed: Post \"https://cert-manager-webhook.cert-manager.svc:443/convert?timeout=30s\": service \"cert-manager-webhook\" not found\n"
```
So doing a delete like you did and reinstalling BKPR fixed my issue.  I still do not understand why it caused such a havoc in our cluster, so I guess it's another thing I'll have to understand before I die!
